### ISSUE TITLE
nan during training

### ISSUE THREAD

**Issue by yhgon** [eriklindernoren/Keras-GAN#87]
Did you see NaN in the training?
configuration
tensorflow : 1.11.0
keras : 2.1.6
GPU : K80 
dataset : apple2orange 

it train well.. but I got NaN in epoch 67. 
 
I've modified the code to use BatchNormalization with  batch(32) instead of Instance Normalization  

gan.train(epochs=200, batch_size=64, sample_interval=15) 

in Epoch 66, I got the result 
```
/usr/local/lib/python3.6/dist-packages/scipy/misc/pilutil.py:482: FutureWarning: Conversion of the second argument of issubdtype from `int` to `np.signedinteger` is deprecated. In future, it will be treated as `np.int64 == np.dtype(int).type`.
  if issubdtype(ts, int):
/usr/local/lib/python3.6/dist-packages/scipy/misc/pilutil.py:485: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  elif issubdtype(type(size), float):
/usr/local/lib/python3.6/dist-packages/keras/engine/training.py:975: UserWarning: Discrepancy between trainable weights and collected trainable weights, did you set `model.trainable` without calling `model.compile` after ?
  'Discrepancy between trainable weights and collected trainable'
[Epoch 0/200] [Batch 0/15] [D loss: 10.340113, acc:  16%] [G loss: 36.504208, adv: 9.286882, recon: 0.813415, id: 1.059065] time: 0:00:46.796445 
[Epoch 0/200] [Batch 1/15] [D loss: 10.501231, acc:  14%] [G loss: 26.232471, adv: 4.194573, recon: 0.792024, id: 1.146999] time: 0:00:51.253706 
[Epoch 0/200] [Batch 2/15] [D loss: 7.842362, acc:  17%] [G loss: 22.555479, adv: 3.907498, recon: 0.631787, id: 1.116496] time: 0:00:55.256927 
[Epoch 0/200] [Batch 3/15] [D loss: 3.473462, acc:  21%] [G loss: 17.249832, adv: 2.483031, recon: 0.506841, id: 1.100872] time: 0:00:59.269601
```
```
[Epoch 67/200] [Batch 7/15] [D loss: 0.071113, acc:  95%] [G loss: 1.885691, adv: 0.042987, recon: 0.077898, id: 0.101703] time: 1:05:19.633317 
[Epoch 67/200] [Batch 8/15] [D loss: 0.096040, acc:  91%] [G loss: 1.872224, adv: 0.072618, recon: 0.075492, id: 0.099593] time: 1:05:23.721828 
[Epoch 67/200] [Batch 9/15] [D loss: 0.151695, acc:  80%] [G loss: 2.235736, adv: 0.179553, recon: 0.082687, id: 0.102930] time: 1:05:27.831364 
[Epoch 67/200] [Batch 10/15] [D loss: 0.140757, acc:  83%] [G loss: 1.851864, adv: 0.066634, recon: 0.074924, id: 0.101068] time: 1:05:31.932562 
[Epoch 67/200] [Batch 11/15] [D loss: 0.093928, acc:  89%] [G loss: 2.103337, adv: 0.081708, recon: 0.085628, id: 0.117779] time: 1:05:36.041804 
[Epoch 67/200] [Batch 12/15] [D loss: nan, acc:  60%] [G loss: 00nan, adv: 00nan, recon: 0.075524, id: 0.119893] time: 1:05:40.103866 
[Epoch 67/200] [Batch 13/15] [D loss: nan, acc:  25%] [G loss: 00nan, adv: 00nan, recon: 00nan, id: 0.092526] time: 1:05:44.104335 
[Epoch 68/200] [Batch 0/15] [D loss: nan, acc:  12%] [G loss: 00nan, adv: 00nan, recon: 00nan, id: 00nan] time: 1:05:48.023691 
/usr/local/lib/python3.6/dist-packages/numpy/core/_methods.py:29: RuntimeWarning: invalid value encountered in reduce
  return umr_minimum(a, axis, None, out, keepdims)
[Epoch 68/200] [Batch 1/15] [D loss: nan, acc:   0%] [G loss: 00nan, adv: 00nan, recon: 00nan, id: 00nan] time: 1:05:52.220977 
[Epoch 68/200] [Batch 2/15] [D loss: nan, acc:   0%] [G loss: 00nan, adv: 00nan, recon: 00nan, id: 00nan] time: 1:05:56.135995 
[Epoch 68/200] [Batch 3/15] [D loss: nan, acc:   0%] [G loss: 00nan, adv: 00nan, recon: 00nan, id: 00nan] time: 1:06:00.045664 
[Epoch 68/200] [Batch 4/15] [D loss: nan, acc:   0%] [G loss: 00nan, adv: 00nan, recon: 00nan, id: 00nan] time: 1:06:03.959619
```

**Comment by yhgon** [comment]
below is my code 
`
from __future__ import print_function, division
import os
import sys

import scipy
import numpy as np

import datetime
import matplotlib.pyplot as plt

from data_loader import DataLoader

from keras.datasets import mnist
#from keras_contrib.layers.normalization import InstanceNormalization
from keras.layers import Input, Dense, Reshape, Flatten, Dropout, Concatenate
from keras.layers import BatchNormalization, Activation, ZeroPadding2D
from keras.layers.advanced_activations import LeakyReLU
from keras.layers.convolutional import UpSampling2D, Conv2D
from keras.models import Sequential, Model
from keras.optimizers import Adam

class CycleGAN():
    def __init__(self):
        # Input shape
        self.img_rows = 128
        self.img_cols = 128
        self.channels = 3
        self.img_shape = (self.img_rows, self.img_cols, self.channels)

        # Configure data loader
        self.dataset_name = 'apple2orange'
        self.data_loader = DataLoader(dataset_name=self.dataset_name,
                                      img_res=(self.img_rows, self.img_cols))


        # Calculate output shape of D (PatchGAN)
        patch = int(self.img_rows / 2**4)
        self.disc_patch = (patch, patch, 1)

        # Number of filters in the first layer of G and D
        self.gf = 32
        self.df = 64

        # Loss weights
        self.lambda_cycle = 10.0                    # Cycle-consistency loss
        self.lambda_id = 0.1 * self.lambda_cycle    # Identity loss

        optimizer = Adam(0.0002, 0.5)

        # Build and compile the discriminators
        self.d_A = self.build_discriminator()
        self.d_B = self.build_discriminator()
        self.d_A.compile(loss='mse',
            optimizer=optimizer,
            metrics=['accuracy'])
        self.d_B.compile(loss='mse',
            optimizer=optimizer,
            metrics=['accuracy'])

        #-------------------------
        # Construct Computational
        #   Graph of Generators
        #-------------------------

        # Build the generators
        self.g_AB = self.build_generator()
        self.g_BA = self.build_generator()

        print('Generator')
        self.g_AB.summary()
        self.g_BA.summary()
        
        # Input images from both domains
        img_A = Input(shape=self.img_shape)
        img_B = Input(shape=self.img_shape)

        # Translate images to the other domain
        fake_B = self.g_AB(img_A)
        fake_A = self.g_BA(img_B)
        # Translate images back to original domain
        reconstr_A = self.g_BA(fake_B)
        reconstr_B = self.g_AB(fake_A)
        # Identity mapping of images
        img_A_id = self.g_BA(img_A)
        img_B_id = self.g_AB(img_B)

        # For the combined model we will only train the generators
        self.d_A.trainable = False
        self.d_B.trainable = False

        # Discriminators determines validity of translated images
        valid_A = self.d_A(fake_A)
        valid_B = self.d_B(fake_B)
        
        print('Discriminators')
        self.d_A.summary()
        self.d_B.summary()

        # Combined model trains generators to fool discriminators
        self.combined = Model(inputs=[img_A, img_B],
                              outputs=[ valid_A, valid_B,
                                        reconstr_A, reconstr_B,
                                        img_A_id, img_B_id ])
        self.combined.summary()
        self.combined.compile(loss=['mse', 'mse',
                                    'mae', 'mae',
                                    'mae', 'mae'],
                            loss_weights=[  1, 1,
                                            self.lambda_cycle, self.lambda_cycle,
                                            self.lambda_id, self.lambda_id ],
                            optimizer=optimizer)

    def build_generator(self):
        """U-Net Generator"""

        def conv2d(layer_input, filters, f_size=4):
            """Layers used during downsampling"""
            d = Conv2D(filters, kernel_size=f_size, strides=2, padding='same')(layer_input)
            d = LeakyReLU(alpha=0.2)(d)
            d = BatchNormalization()(d) #d = InstanceNormalization()(d)  
            return d

        def deconv2d(layer_input, skip_input, filters, f_size=4, dropout_rate=0):
            """Layers used during upsampling"""
            u = UpSampling2D(size=2)(layer_input)
            u = Conv2D(filters, kernel_size=f_size, strides=1, padding='same', activation='relu')(u)
            if dropout_rate:
                u = Dropout(dropout_rate)(u)
            u = BatchNormalization()(u) # u = InstanceNormalization()(u)
            u = Concatenate()([u, skip_input])
            return u

        # Image input
        d0 = Input(shape=self.img_shape)

        # Downsampling
        d1 = conv2d(d0, self.gf)
        d2 = conv2d(d1, self.gf*2)
        d3 = conv2d(d2, self.gf*4)
        d4 = conv2d(d3, self.gf*8)

        # Upsampling
        u1 = deconv2d(d4, d3, self.gf*4)
        u2 = deconv2d(u1, d2, self.gf*2)
        u3 = deconv2d(u2, d1, self.gf)

        u4 = UpSampling2D(size=2)(u3)
        output_img = Conv2D(self.channels, kernel_size=4, strides=1, padding='same', activation='tanh')(u4)

        return Model(d0, output_img)

    def build_discriminator(self):

        def d_layer(layer_input, filters, f_size=4, normalization=True):
            """Discriminator layer"""
            d = Conv2D(filters, kernel_size=f_size, strides=2, padding='same')(layer_input)
            d = LeakyReLU(alpha=0.2)(d)
            if normalization:
                d = BatchNormalization()(d)  # d = InstanceNormalization()(d)
            return d

        img = Input(shape=self.img_shape)

        d1 = d_layer(img, self.df, normalization=False)
        d2 = d_layer(d1, self.df*2)
        d3 = d_layer(d2, self.df*4)
        d4 = d_layer(d3, self.df*8)

        validity = Conv2D(1, kernel_size=4, strides=1, padding='same')(d4)

        return Model(img, validity)

    def train(self, epochs, batch_size=1, sample_interval=50):

        start_time = datetime.datetime.now()

        # Adversarial loss ground truths
        valid = np.ones((batch_size,) + self.disc_patch)
        fake = np.zeros((batch_size,) + self.disc_patch)

        for epoch in range(epochs):
            for batch_i, (imgs_A, imgs_B) in enumerate(self.data_loader.load_batch(batch_size)):

                # ----------------------
                #  Train Discriminators
                # ----------------------

                # Translate images to opposite domain
                fake_B = self.g_AB.predict(imgs_A)
                fake_A = self.g_BA.predict(imgs_B)

                # Train the discriminators (original images = real / translated = Fake)
                dA_loss_real = self.d_A.train_on_batch(imgs_A, valid)
                dA_loss_fake = self.d_A.train_on_batch(fake_A, fake)
                dA_loss = 0.5 * np.add(dA_loss_real, dA_loss_fake)

                dB_loss_real = self.d_B.train_on_batch(imgs_B, valid)
                dB_loss_fake = self.d_B.train_on_batch(fake_B, fake)
                dB_loss = 0.5 * np.add(dB_loss_real, dB_loss_fake)

                # Total disciminator loss
                d_loss = 0.5 * np.add(dA_loss, dB_loss)


                # ------------------
                #  Train Generators
                # ------------------

                # Train the generators
                g_loss = self.combined.train_on_batch([imgs_A, imgs_B],
                                                        [valid, valid,
                                                        imgs_A, imgs_B,
                                                        imgs_A, imgs_B])

                elapsed_time = datetime.datetime.now() - start_time

                # Plot the progress
                print ("[Epoch %d/%d] [Batch %d/%d] [D loss: %f, acc: %3d%%] [G loss: %05f, adv: %05f, recon: %05f, id: %05f] time: %s " \
                                                                        % ( epoch, epochs,
                                                                            batch_i, self.data_loader.n_batches,
                                                                            d_loss[0], 100*d_loss[1],
                                                                            g_loss[0],
                                                                            np.mean(g_loss[1:3]),
                                                                            np.mean(g_loss[3:5]),
                                                                            np.mean(g_loss[5:6]),
                                                                            elapsed_time))

                # If at save interval => save generated image samples
                if batch_i % sample_interval == 0:
                    self.sample_images(epoch, batch_i)

    def sample_images(self, epoch, batch_i):
        os.makedirs('images/%s' % self.dataset_name, exist_ok=True)
        r, c = 2, 3

        imgs_A = self.data_loader.load_data(domain="A", batch_size=1, is_testing=True)
        imgs_B = self.data_loader.load_data(domain="B", batch_size=1, is_testing=True)

        # Demo (for GIF)
        #imgs_A = self.data_loader.load_img('datasets/apple2orange/testA/n07740461_1541.jpg')
        #imgs_B = self.data_loader.load_img('datasets/apple2orange/testB/n07749192_4241.jpg')

        # Translate images to the other domain
        fake_B = self.g_AB.predict(imgs_A)
        fake_A = self.g_BA.predict(imgs_B)
        # Translate back to original domain
        reconstr_A = self.g_BA.predict(fake_B)
        reconstr_B = self.g_AB.predict(fake_A)

        gen_imgs = np.concatenate([imgs_A, fake_B, reconstr_A, imgs_B, fake_A, reconstr_B])

        # Rescale images 0 - 1
        gen_imgs = 0.5 * gen_imgs + 0.5

        titles = ['Original', 'Translated', 'Reconstructed']
        fig, axs = plt.subplots(r, c)
        cnt = 0
        for i in range(r):
            for j in range(c):
                axs[i,j].imshow(gen_imgs[cnt])
                axs[i, j].set_title(titles[j])
                axs[i,j].axis('off')
                cnt += 1
        fig.savefig("images/%s/cyclegan_%d_%d.png" % (self.dataset_name, epoch, batch_i))
        plt.close()


`

**Comment by Feiyu-Zhang** [comment]
> I've modified the code to use BatchNormalization with batch(32) instead of Instance Normalization
> 
> below is my code the difference for generator.
> 
> ```
>             d = BatchNormalization()(d) #d = InstanceNormalization()(d)  
>             u = BatchNormalization()(u) # u = InstanceNormalization()(u)
> ```
> 
> for discriminator
> 
> ```
>             if normalization:
>                 d = BatchNormalization()(d)  # d = InstanceNormalization()(d)
>             return d
> ```

Please have you found the reason?
